### PR TITLE
[DEV-1876] Alchemy

### PR DESCRIPTION
**Pull request by bernd2022**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations
  - Login/logout
  - Registration
  - Update user data
  - Create/delete payment routes
- [ ] **Update environment**

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

File: src/config/config.ts
Comment (davidleomay) [inline src/config/config.ts:1]: Scan API config can be removed for Alchemy chains

File: src/integration/alchemy/services/alchemy-webhook.service.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    const network = AlchemyNetworkMapper.toAlchemyNetworkByBlockchain(dto.blockchain);
+    if (!network) return;
+
+    const allWebhooks = await this.alchemy.notify.getAllWebhooks();
+    await Promise.all(
+      allWebhooks.webhooks.filter((wh) => wh.network === network).map((wh) => this.alchemy.notify.deleteWebhook(wh.id)),
+    );
+
+    //const url = `${Config.url}/alchemy/addressWebhook`;
+    const url = 'https://dev.api.dfx.swiss/v1/alchemy/addressWebhook';
```
Comment (davidleomay) [inline src/integration/alchemy/services/alchemy-webhook.service.ts:54]: Remove comment, use URL from config

File: src/subdomains/core/liquidity-management/adapters/balances/blockchain.adapter.ts [hunk 2]
```diff
@@ -154,6 +157,34 @@ export class BlockchainAdapter implements LiquidityBalanceIntegration {
...
+        const balance =
+          asset.type === AssetType.COIN
+            ? client.fromWeiAmount(coinBalance)
+            : tokenToBalanceMap.get(asset.dexName.toUpperCase()) ?? 0;
+
+        this.balanceCache.set(asset.id, balance);
+      } catch (e) {
+        this.logger.error(`Failed to update liquidity management balance for ${asset.uniqueName}:`, e);
+        this.invalidateCacheFor([asset]);
+      }
```
Comment (davidleomay) [inline src/subdomains/core/liquidity-management/adapters/balances/blockchain.adapter.ts:178]: The try-catch doesn't make much sense here. Should be around lines 160-161 (and invalidate cache for all assets).

File: src/subdomains/supporting/payin/strategies/register/impl/arbitrum.strategy.ts [hunk 3]
```diff
@@ -24,6 +24,14 @@ export class ArbitrumStrategy extends EvmStrategy {
     super('ETH', arbitrumService, payInRepository, assetService, repos);
   }
 
+  onModuleInit() {
+    this.addressWebhookMessageQueue = new QueueHandler();
+
+    this.alchemyWebhookService
+      .getAddressWebhookObservable([Network.ARB_MAINNET, Network.ARB_GOERLI])
+      .subscribe((dto) => this.processAddressWebhookMessageQueue(dto));
```
Comment (davidleomay) [inline src/subdomains/supporting/payin/strategies/register/impl/arbitrum.strategy.ts:32]: Use the AlchemyNetworkMapper (for `this.blockchain`)? Same for other chains.

File: src/subdomains/supporting/payin/strategies/register/impl/arbitrum.strategy.ts [hunk 4]
```diff
@@ -66,11 +74,8 @@ export class ArbitrumStrategy extends EvmStrategy {
...
   //*** JOBS ***//
 
-  @Cron(CronExpression.EVERY_MINUTE)
-  @Lock(7200)
   async checkPayInEntries(): Promise<void> {
-    if (Config.processDisabled(Process.PAY_IN)) return;
-
+    this.logger.info('Arbitrum checkPayInEntries() ...');
     await this.processNewPayInEntries();
   }
```
Comment (davidleomay) [inline src/subdomains/supporting/payin/strategies/register/impl/arbitrum.strategy.ts:80]: Remove (also for other chains)?

File: src/subdomains/supporting/payin/strategies/register/impl/base/evm.strategy.ts [hunk 5]
```diff
@@ -165,4 +174,68 @@ export abstract class EvmStrategy extends RegisterStrategy {
...
+    toAddresses: string[],
+    dto: AlchemyWebhookDto,
+  ): AlchemyWebhookActivityDto[] {
+    const notFromOwnAddresses = dto.event.activity.filter(
+      (tx) => !fromAddresses.map((a) => a.toLowerCase()).includes(tx.fromAddress.toLowerCase()),
+    );
+
+    return notFromOwnAddresses.filter((tx) =>
+      toAddresses.map((a) => a.toLowerCase()).includes(tx.toAddress.toLowerCase()),
+    );
```
Comment (davidleomay) [inline src/subdomains/supporting/payin/strategies/register/impl/base/evm.strategy.ts:211]: Is this separated on purpose? Could be merged.

File: src/integration/blockchain/shared/evm/evm-client.ts [hunk 6]
```diff
@@ -339,38 +378,94 @@ export abstract class EvmClient {
...
+
+    while (pageKey) {
+      assetTransfersResponse = await this.alchemyGetAssetTransfers(params, pageKey);
+      pageKey = assetTransfersResponse.pageKey;
+
+      assetTransferResult.push(...assetTransfersResponse.transfers);
+    }
+
+    return assetTransferResult;
+  }
```
Comment (davidleomay) [inline src/integration/blockchain/shared/evm/evm-client.ts:429]: You could write a pagination helper method and reuse it in `getTokenBalancesByAlchemy` 😉

File: src/integration/alchemy/controllers/alchemy.controller.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+  @Get('webhookAddresses/:id')
+  async getWebhookAddresses(@Param('id') id: string): Promise<AddressActivityResponse> {
+    return this.alchemyWebhookService.getWebhookAddresses(id);
+  }
+
+  @Post('createAddressWebhook')
+  async createAddressWebhook(@Body() dto: CreateWebhookDto) {
+    return this.alchemyWebhookService.createAddressWebhook(dto);
+  }
```
[the remark this answers is not in the document]
Comment (bernd2022) [reply src/integration/alchemy/controllers/alchemy.controller.ts:29]: No, they where just be provided for testing purposes

File: src/subdomains/supporting/payin/strategies/register/impl/base/evm.strategy.ts (hunk 5 above)
Comment (bernd2022) [reply src/subdomains/supporting/payin/strategies/register/impl/base/evm.strategy.ts:211]: It's the same in line 134-138. In my opinion it is helpful in this case to have the variable "notFromOwnAddresses" as a kind of documentation, what happens here.

File: src/subdomains/core/liquidity-management/adapters/balances/blockchain.adapter.ts [hunk 8]
```diff
@@ -154,6 +158,38 @@ export class BlockchainAdapter implements LiquidityBalanceIntegration {
...
+    let tokenBalances: OwnedToken[];
+
+    try {
+      coinBalance = await client.getNativeCoinBalanceByAlchemy();
+      tokenBalances = await client.getTokenBalancesByAlchemy();
+    } catch (e) {
+      this.logger.error('Failed to update liquidity management balance for all assets:', e);
+      this.invalidateCacheFor(assets);
+      return;
+    }
```
Comment (davidleomay) [inline src/subdomains/core/liquidity-management/adapters/balances/blockchain.adapter.ts:171]: Add the blockchain to the log

File: src/config/config.ts [hunk 9]
```diff
@@ -292,8 +292,6 @@ export class Configuration {
       }),
     },
     ethereum: {
-      ethScanApiUrl: process.env.ETH_SCAN_API_URL,
-      ethScanApiKey: process.env.ETH_SCAN_API_KEY,
```
Comment (davidleomay) [inline src/config/config.ts:296]: Remove in Bicep (let's do it together 😉)

Comment (bernd2022) [reply src/config/config.ts:296]: let's do it together is a great idea 😉

**Review by davidleomay — APPROVED** [review]

**Review by davidleomay — APPROVED** [review]

**Review by davidleomay — CHANGES_REQUESTED** [review]
Update Bicep with Alchemy settings

**Pull request merged**